baldur/src/main/java/org/opencadc/baldur/GetAction.java

```diff
@@ -0,0 +1,241 @@
+        
+        if (op == null) {
+            throw new IllegalArgumentException("missing required parameter, " + OP);
+        }
+        Operation operation = null;
+        try {
+            operation = Operation.valueOf(op);
+        } catch (IllegalArgumentException e) {
+            throw new IllegalArgumentException("invalid " + OP + " parameter, must be "
+                + Operation.read  + " or " + Operation.write);
+        }
+        if (uri == null) {
+            throw new IllegalArgumentException("missing required parameter, " + URI);
+        }
+        
+        URI artifactURI;
+        try {
+            artifactURI = new URI(uri);
+            InventoryUtil.validateArtifactURI(GetAction.class, artifactURI);
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("invalid " + URI + " parameter, not a valid URI: " + uri);
+        }
+
+        Grant grant = null;
+        switch (operation) {
+            case read:
+                grant = getReadGrant(permissionsConfig, artifactURI);
+                break;
+            case write:
+                grant = getWriteGrant(permissionsConfig, artifactURI);
+                break;
+            default:
+                throw new IllegalStateException("unknown operation: " + operation);
+        }
+
+        syncOutput.setHeader("Content-Type", "text/xml");
+        syncOutput.setCode(200);
+
+        OutputStream out = syncOutput.getOutputStream();
+        GrantWriter writer = new GrantWriter();
+        writer.write(grant, out);
+        out.flush();
+
+    }
+
+    /**
+     * Calculate the expiry date of the grant for the given artifact URI.
+     *
+     * @return A Date.
+     */
+    private static Date getExpiryDate() {
+        // expires immediately
+        return new Date();
+    }
+    
+    /**
+     * Get the read grant for the given Artifact URI. 
+     *
+     * @param artifactURI The Artifact URI.
+     * @return The read grant information object for the artifact URI.
+     */
+    static ReadGrant getReadGrant(PermissionsConfig permissionsConfig, URI artifactURI) {
+        InventoryUtil.assertNotNull(PermissionsConfig.class, "artifactURI", artifactURI);
+        boolean anonymousRead = false;
+        List<GroupURI> groups = new ArrayList<GroupURI>();
+        Iterator<PermissionEntry> matchingEntries = permissionsConfig.getMatchingEntries(artifactURI);
+        PermissionEntry next = null;
+        log.debug("compiling read grant from matching entries");
+        while (matchingEntries.hasNext()) {
+            next = matchingEntries.next();
+            log.debug("matching entry: " + next);
+            if (!anonymousRead) {
+                anonymousRead = next.anonRead;
+            }
+            for (GroupURI groupURI : next.readOnlyGroups) {
+                if (!groups.contains(groupURI)) {
+                    groups.add(groupURI);
+                }
+            }
+        }
+        
```

Comment:
This needs to also loop over readWrite groups as they also have read permission

Reply:
right, of course